tests/test_dmlscheduler.py

```diff
@@ -0,0 +1,93 @@
+import pytest
+import time
+import tests.context
+
+import logging
+import numpy as np
+from custom.keras               import get_optimizer
+from core.runner                import DMLRunner
+from core.runner_actor          import DMLRunnerActor
+from models.keras_perceptron    import KerasPerceptron
+from core.utils.dmljob          import DMLJob, serialize_job, deserialize_job
+from core.scheduler             import DMLScheduler
+from examples.labelers          import mnist_labeler
+from queue import Queue
+print("testing")
+
+
+import ray
+
+
+def make_dataset_path():
+    return 'datasets/mnist'
+
+
+def make_config():
+    return {}
+
+
+def make_model_json():
+    m = KerasPerceptron(is_training=True)
+    model_architecture = m.model.to_json()
+    model_optimizer = get_optimizer(m.model)
+    model_json = {
+        "architecture": model_architecture,
+        "optimizer": model_optimizer
+    }
+    return model_json
+
+
+def make_initialize_job(model_json):
+    initialize_job = DMLJob(
+        "initialize",
+        model_json,
+        "keras"
+    )
+    return initialize_job
+
+scheduler = DMLScheduler()
+
+#Check that the scheduling/running functionality is maintained.
+def test_dmlscheduler_sanity():
+    model_json = make_model_json()
+    initialize_job = make_initialize_job(model_json)
+    scheduler.add_job(initialize_job)
+    scheduler._runners_run_next_jobs()
+    start = time.time()
+    while not scheduler.processed:
+        scheduler._runners_run_next_jobs()
+    initial_weights = scheduler.processed.pop(0)
+    print(time.time()-start)
+    assert type(initial_weights) == list
+    assert type(initial_weights[0]) == np.ndarray
+    scheduler.reset()
+
+#Check that running 10 jobs with Ray takes at most half the 
+# time to run them serially (i.e. there is a speedup)
+def test_dmlscheduler_speedup_naive():
+    scheduler.reset()
+    slow = 0
+    for _ in range(10):
+        model_json = make_model_json()
+        initialize_job = make_initialize_job(model_json)
+        scheduler.add_job(initialize_job)
+        start = time.time()
+        while len(scheduler.processed) < 1:
+            scheduler._runners_run_next_jobs()
+        slow += time.time() - start
+        scheduler.reset()
+
+    for _ in range(10):
+        model_json = make_model_json()
+        initialize_job = make_initialize_job(model_json)
+        scheduler.add_job(initialize_job)
+    start = time.time()
+    while len(scheduler.processed) < 10:
+        scheduler._runners_run_next_jobs()
+    fast = time.time() - start
+    print(slow)
+    print(fast)
+    assert fast*1.5 < slow
+    scheduler.reset()
+
+
```

Comment:
Can we add a test that checks if 5 or 10 things can be scheduled at different times (for example with five things: scheduling 2 at the beginning, 2 after the 1st executes, one after the 4th executes) and verify that all jobs were executed? 

Also a test that checks if jobs are picked up and executed regardless of the scheduler being interacted with (basically that line 81 in the `scheduler.py` works)? (This would need to run fast, so the `run_frequency_in_mins` should be set to something like 5 seconds to keep the whole test under 10 seconds.)

Reply:
The first test sounds good, I'll implement that.

As for the second test: since the CRON job method is something you wrote, I had a quick question. When the CRON job is set up, does it asynchronously call the method I specified every period (so the proper way to test would be to wait a few periods and then check if all jobs are done running)?

Reply:
Ok. Once the first test is implemented and the minor things are changed we will be ready to merge.

For the second one, can you check [the documentation for the `schedule` module](http://schedule.readthedocs.io/en/stable/)? I found the module by googling but I wasn't able to test it until the multithreading capabilities were in place. I think you're correct on your approach to test it. We're not locked into the module, so if it's too hard to set up you can change it. We just want things to run as long as there are things to be run (as long as things are in the queue).

Reply:
Completed both tests. 

Messaged you on Slack regarding the issues of `schedule` and `ray` altogether in the process of making the second test, and why I ended up using `multiprocessing`